### DISCUSSION TITLE
Unrecognized Inverter - Sofar

### DISCUSSION THREAD

**Discussion by wills106** [wills106/homeassistant-solax-modbus#522] [Q&A]
If you get the following error in your log:
`"unrecognized inverter type - serial number : {your_serial_number_here}"`

Please list below the first 6 digits here including the Inverter type and kW rating and which plugin you got a response with.
I can then add in your Inverter.
**Please do not raise an issue in addition.**

If you get `unrecognized XYZ inverter type - serial number : unknown` you don't have a working Modbus connection.
**Don't raise an issue, this isn't a fault of the Integration.**
Either use one of the existing discussions or start a new one, don't comment in this discussion about `unrecognized XYZ inverter type - serial number : unknown`

**Comment by RadovanMicek** [comment]
Hello does anyone solve this? How to setup please? Thank you :-) 
HYD 10KTL-3PH-PRO
SP2ES1XXXXXXXX

**Reply by DennyRuza** [reply]
Hi @RadovanMicek,

were you able to solve this?
I have the same inverter, I am able to change the parameters such as Energy Storage Mode, FeedIn: Limitation Mode, etc. via Homeassistant, however all of the sensors have the unknown status.
@wills106 Thank you for all your work. Would it be possible to add our inverter to the next update?

**Comment by RadovanMicek** [comment]
2023-09-08 14:10:00.550 ERROR (MainThread) [pymodbus.logging] Connection to (192.168.0.101, 502) failed: [Errno 111] Connection refused
2023-09-08 14:10:00.564 WARNING (MainThread) [custom_components.solax_modbus.plugin_sofar] sofar: attempt to read serialnumber failed at 0x445
Traceback (most recent call last):
  File "/config/custom_components/solax_modbus/plugin_sofar.py", line 56, in _read_serialnr
    inverter_data = hub.read_holding_registers(unit=hub._modbus_addr, address=address, count=7)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/solax_modbus/__init__.py", line 336, in read_holding_registers
    return self._client.read_holding_registers(address, count, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pymodbus/client/mixin.py", line 102, in read_holding_registers
    return self.execute(
           ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pymodbus/client/base.py", line 175, in execute
    raise ConnectionException(f"Failed to connect[{str(self)}]")
pymodbus.exceptions.ConnectionException: Modbus Error: [Connection] Failed to connect[ModbusTcpClient(192.168.0.101:502)]
2023-09-08 14:10:00.570 WARNING (MainThread) [custom_components.solax_modbus.plugin_sofar] sofar: reading serial number from address 0x445 failed; other address may succeed
2023-09-08 14:10:00.572 ERROR (MainThread) [custom_components.solax_modbus.plugin_sofar] sofar: cannot find serial number, even not for other Inverter
2023-09-08 14:10:00.573 ERROR (MainThread) [custom_components.solax_modbus.plugin_sofar] unrecognized sofar inverter type - serial number : unknown

**Comment by jordisco** [comment]
I have a Sofar 100KTL 100 Kw, the serial starts with SQ1ES1, the error is "unrecognized solarx inverter type - serial number: SQ1ES1xxxxxxxx"

**Reply by wills106** [reply]
Is this a PV Only 3 Phase Inverter?
Was that response with plugin_sofar or plugin_sofar_old?

**Reply by jordisco** [reply]
yes tree phase inverter.
with plugin_sofa

**Reply by wills106** [reply]
I have just released [2023.09.7](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2023.09.7) I have only added in MPPT 1-4 so far and you need to manually enable Strings 3 & 4.
Can you let me know if PV Power Total is scaled correctly?

I'll looking at adding in the extra PV Strings. Are you using all 10?

**Reply by jordisco** [reply]
ok, it seems to work correctly.

<img width="315" alt="Captura de pantalla 2023-09-26 a las 12 40 47" src="https://github.com/wills106/homeassistant-solax-modbus/assets/68939441/35a270f5-25e2-4021-8d5f-9d25c9fca957">
<img width="315" alt="Captura de pantalla 2023-09-26 a las 12 41 11" src="https://github.com/wills106/homeassistant-solax-modbus/assets/68939441/21548aa1-d63b-445a-9d1b-b3b83863d78f">

<img width="318" alt="Captura de pantalla 2023-09-26 a las 12 41 34" src="https://github.com/wills106/homeassistant-solax-modbus/assets/68939441/077213df-257b-44fb-91aa-17dcec9e9312">

**Reply by jordisco** [reply]
export and import sensors do not work

**Comment by robrankin** [comment]
HYD-5000-ES
SM1ES1xxxx
Using the sofar_old plugin.
Component Version: 2023.09.7
HA Version: 2023.9.3


Far as I can tell this should be a recognized inverter type, as the plugin should match the serial here: https://github.com/wills106/homeassistant-solax-modbus/blob/ebf5a81fb93f0bef5b2a8f61679e77ea576d4be6/custom_components/solax_modbus/plugin_sofar_old.py#L1080C41-L1080C41

However, it just.. doesn't....

![error](https://github.com/wills106/homeassistant-solax-modbus/assets/1881181/54412ac2-a973-4b1d-881d-a70f07e24c16)

**Reply by robrankin** [reply]
I've swapped cabling between the Waveshare RS485/TCP converter and have also upgraded the inverters firmware, no change.

Digging into the logs a bit I've noticed this:

![2023-09-30 11_23_23-Clipboard](https://github.com/wills106/homeassistant-solax-modbus/assets/1881181/4eceeda2-444e-4c38-9d2a-d1c2fc7059f8)

It appears the serial number is being prepended with some control characters.  Not sure, but that seems possibly the problem.  Any ideas?

**Reply by robrankin** [reply]
Changed the Waveshare converter for a USB RS485 and mbusd on a RPi located with the inverter, no change.

**Reply by wills106** [reply]
If you change 0x2002 to 0x2003

https://github.com/wills106/homeassistant-solax-modbus/blob/ebf5a81fb93f0bef5b2a8f61679e77ea576d4be6/custom_components/solax_modbus/plugin_sofar_old.py#L1062-L1065

Does it just read the SM1E

**Reply by robrankin** [reply]
Seems not, seeing the same data in the logs.

![2023-10-02 14_26_47-Clipboard](https://github.com/wills106/homeassistant-solax-modbus/assets/1881181/d6eb0fec-dfaf-4684-a780-59a0dbcb155e)

**Reply by wills106** [reply]
Have you double checked the baud rate matches on the Inverter and Waveshare? Default Sofar is 9600. I would imagine you can check in the Inverter LCD.

Might be worth also moving the 0x2002 to 0x2004. It's hard to tell how many rubbish characters are in front of SM1E

**Reply by robrankin** [reply]
Same problem with 0x2004.

I've tested with both the Waveshare device and mbusd running on an rpi, same result.  I would assume I wouldn't even see a partially correct serial if the baud rate didnt match?

**Reply by robrankin** [reply]
Just to test, I changed:

```
        elif seriesnumber.startswith('SM1E'):  invertertype = HYBRID | X1 # 3kW HYDxxxxES
```

To:

```
        elif seriesnumber.count('SM1E'):  invertertype = HYBRID | X1 # 3kW HYDxxxxES
```

Now the plugin returns 38 entities, that seem to be reading correctly:

![image](https://github.com/wills106/homeassistant-solax-modbus/assets/1881181/1ec42bca-f657-4328-b36e-abd141d1d40d)

**Reply by wills106** [reply]
I have just released [2023.10.1](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2023.10.1) adding in this change.

**Comment by ggtimtom** [comment]
Please list below the first 6 digits here 
**ZS2ES012**
including the Inverter type and 
**Azzurro 3PH 12KTL-V3**

kW rating and 
**12kW**

which plugin you got a response with.
**sofar**

I can then add in your Inverter.
**Thank you!**

**Reply by ggtimtom** [reply]
https://www.zcsazzurro.com/de/inverter/azzurro-three-phase-string-inverter-compact-series-3-3-12-kw-v3

**Reply by ggtimtom** [reply]
For testing i made a fork and edited this line:
elif seriesnumber.startswith('SS2E'):  invertertype = PV | X3 | GEN # 4.4 KTLX-G3
to this:
elif seriesnumber.startswith('ZS2E'):  invertertype = PV | X3 | GEN # 4.4 KTLX-G3

And it works fine.

**Reply by wills106** [reply]
I have just released [2023.10.1](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2023.10.1) adding in your Serial.

**Comment by Maximimimi** [comment]
Hello I just managed to connect my Inverter.

- Inverter @ 15kW :
**Azzurro 3PH HYD15000 ZSS**
- Serial Number:
**ZP1ES0xxxxxxxx**
- Plugin I got a response with:
**sofar**

I tested changing the line : 
elif seriesnumber.startswith('SP1'):  invertertype = HYBRID | X3 | GEN # HYDxxKTL-3P
to
elif seriesnumber.startswith('ZP1'):  invertertype = HYBRID | X3 | GEN # HYDxxKTL-3P
And it works.

Would be great if you could add the Inverter.
Thank you

**Reply by wills106** [reply]
Sorry I missed this post. Added into [2023.11.1](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2023.11.1)

**Comment by andbad** [comment]
Hi,
I tryied to connect my inverter but not recognized

rated power: **8kW**
Model: Azzurro **3PH HYD8000 ZSS**
Serial Number: **ZP2ES0xxxxxxxx**

Plugin I got a response with:
sofar

By(t)e

**Reply by andbad** [reply]
I added this line:
`        elif seriesnumber.startswith('ZP2'):  invertertype = HYBRID | X3 | GEN # Azzurro HYDxx ZSS`
And it works

By(t)e

**Reply by wills106** [reply]
Thanks, I'll add it into the next release.

**Reply by andbad** [reply]
How can I find if it's a GEN3 or GEN4?

By(t)e

**Reply by andbad** [reply]
I tried with GEN4 but not works, so I tink it's a GEN3
Unfortunally, I can't change "User mode", in log I see:
```
2024-02-14 12:56:17.091 ERROR (SyncWorker_13) [pymodbus.logging] General exception: unpack requires a buffer of 4 bytes
2024-02-14 12:56:17.091 ERROR (SyncWorker_13) [pymodbus.logging] Modbus IO exception Modbus Error: [Input/Output] Unable to decode request

```

Have you some hints?

By(t)e

**Comment by weemaba999** [comment]
Please list below the first 6 digits here
SE1ES4
including the Inverter type and
ME3000SP

kW rating and
3

which plugin you got a response with.
sofar

Modbus adapter:
LSE-3 solar ethernet logger

I can then add in your Inverter.
Thank you!

**Reply by wills106** [reply]
Does pluging_sofar return your Serial? I thought the ME3000SP used plugin_sofar_old

**Reply by weemaba999** [reply]
Hi;

Thanks a lot for your answer. The plugin_sofar_old returns all my values & serial. Now, is it possible to write to the inverter to work with the passive mode ? I have no possibility in the discovered entities to change something or to change normal to passive mode.
(Maybe I should first manually change to passive mode before entities are visible to 'steer' the inverter) ?

Kr,

Bart

**Reply by weemaba999** [reply]
@wills106 can you have a look at my question please ? I'll leave you alone after this ;-)

**Reply by wills106** [reply]
I haven't added in any control on sofar_old as little feedback from users with Inverters using that register set.
Do you want to raise a Feature Request under the Issues section and we can track progress there.

**Comment by Sebmad** [comment]
Hello I just managed to connect my Inverter.

- Inverter @ 6kW :
     **Sofar Solar ESI 6K-S1**

- Serial Number:
    **SQ1025***************

- Plugin I got a response with:
    **sofar**
I tested changing the line :
  seriesnumber.startswith("SQ1"): invertertype = HYBRID | X1 | GEN  # HYDxxxxKTL ZCS HP, Single Phase
And it works.

Would be great if you could add the Inverter.
Thank you

**Reply by ed-fr** [reply]
Hello, 
Same problem, same solution

Inverter @ 6kW :
Sofar Solar ESI 6K-S1

Serial Number:
SQ1025*************

Plugin I got a response with:
sofar
I tested changing the line :
seriesnumber.startswith("SQ1"): invertertype = HYBRID | X1 | GEN # HYDxxxxKTL ZCS HP, Single Phase

It works, but there are some inconsistencies and features that don't function. I've attached the MODBUS information from SOFAR SOLAR in case it helps...

[SOFAR Modbus User Guide 2023-07-23.pdf](https://github.com/user-attachments/files/23601083/SOFAR.Modbus.User.Guide.2023-07-23.pdf)

[Sofar_Inverter_MODBUS_V1.35_EN.pdf](https://github.com/user-attachments/files/23601086/Sofar_Inverter_MODBUS_V1.35_EN.pdf)

**Reply by ed-fr** [reply]
I can confirm that the feedback function is working, but the actions are not.
